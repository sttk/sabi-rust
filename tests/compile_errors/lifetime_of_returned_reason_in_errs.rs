use sabi::Err;

#[derive(Debug)]
enum Reasons {
  MyErr { msg: String, flag: bool },
}

fn return_reason() -> Result<(), Reasons> {
  Err(Reasons::MyErr { msg: "hello".to_string(), flag: true })
}

fn forward_reason() -> Result<(), Reasons> {
  return_reason()
}

fn return_err() -> Result<(), Err> {
  let err = Err::new(Reasons::MyErr { msg: "hello".to_string(), flag: true });
  Err(err)
}

fn forward_reason_in_err() -> Result<&'static Reasons, &'static Err> {
  let err = return_err().unwrap_err();
  err.reason::<Reasons>()
}

fn main() {
  let reason = forward_reason().unwrap_err();
  println!("{:?}", reason);

  let reason = forward_reason_in_err().unwrap_err();
  println!("{:?}", reason);
}
