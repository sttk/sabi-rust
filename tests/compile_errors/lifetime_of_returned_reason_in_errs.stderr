error[E0515]: cannot return value referencing local variable `err`
  --> tests/compile_errors/lifetime_of_returned_reason_in_errs.rs:23:3
   |
23 |   err.reason::<Reasons>()
   |   ---^^^^^^^^^^^^^^^^^^^^
   |   |
   |   returns a value referencing data owned by the current function
   |   `err` is borrowed here
